common/task/ackMgr.go

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package task
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/uber/cadence/common/collection"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/metrics"
+)
+
+type (
+	// AckMgr is the interface for reading/acknowledging Tasks
+	AckMgr interface {
+		ReadTasks() ([]Info, bool, error)
+		CompleteTask(SequenceID)
+		AckLevel() SequenceID
+		ReadLevel() SequenceID
+		UpdateAckLevel() SequenceID
+		PendingTasks() int64
+	}
+
+	// AckMgrBase is the interface for the functionality that AckMgr relies on
+	AckMgrBase interface {
+		PaginationFn(SequenceID, SequenceID) collection.PaginationFn
+		UpdateMaxReadLevel() SequenceID
+	}
```

Comment:
this interface may not be necessary, since each replication / transfer / timer component (event loop) will do the IO loading task then adding new task to ack manager